salt/proxy/restconf.py

```diff
@@ -0,0 +1,222 @@
+
+password:
+    The password to pass to the device to authenticate the RESTCONF requests.
+
+# TODO: timeout not yet implemented
+# timeout: ``60``
+#     Time in seconds to wait for the device to respond. Default: 60 seconds.
+
+verify: ``True``
+    Either a boolean, in which case it controls whether we verify the NX-API
+    TLS certificate, or a string, in which case it must be a path to a CA bundle
+    to use. Defaults to ``True``.
+
+    When there is no certificate configuration on the device and this option is
+    set as ``True`` (default), the commands will fail with the following error:
+    ``SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:581)``.
+    In this case, you either need to configure a proper certificate on the
+    device (*recommended*), or bypass the checks setting this argument as ``False``
+    with all the security risks considered as you may be MITM'd.
+
+Proxy Pillar Example
+--------------------
+
+.. code-block:: yaml
+
+    proxy:
+      proxytype: restconf
+      host: switch1.example.com
+      username: example
+      password: example
+      verify: false
+"""
+
+# Import python stdlib
+import copy
+import json
+import logging
+
+import salt.utils.http
+
+# Import Salt modules
+from salt.exceptions import SaltException
+
+# -----------------------------------------------------------------------------
+# proxy properties
+# -----------------------------------------------------------------------------
+
+__proxyenabled__ = ["restconf"]
+# proxy name
+
+# -----------------------------------------------------------------------------
+# globals
+# -----------------------------------------------------------------------------
+
+__virtualname__ = "restconf"
+log = logging.getLogger(__file__)
+restconf_device = {}
+
+# -----------------------------------------------------------------------------
+# property functions
+# -----------------------------------------------------------------------------
+
+
+def __virtual__():
+    """
+    This Proxy Module is widely available as there are no external dependencies.
+    """
+    log.debug("restconf proxy __virtual__() called...")
+    return __virtualname__
+
+
+# -----------------------------------------------------------------------------
+# proxy functions
+# -----------------------------------------------------------------------------
+
+
+def init(opts):
+    log.debug("restconf proxy init(opts) called...")
+    # Open the connection to the RESTCONF Device.
+    # As the communication is HTTP based, there is no connection to maintain,
+    # however, in order to test the connectivity and make sure we are able to
+    # bring up this Minion, we are checking the standard restconf state uri.
+
+    proxy_dict = opts.get("proxy", {})
+    conn_args = copy.deepcopy(proxy_dict)
+    conn_args.pop("proxytype", None)
+    opts["multiprocessing"] = conn_args.pop("multiprocessing", True)
```

Comment:
Are these `opts` just limited to this? If this is changing stuff in the broad salt multiprocessing opts, I know there be dragons :dragon:

Reply:
Its on purpose to make this single minion-proxy process use multiprocessing. I took this part from the napalm & nxos implementation which is doing this in the same way